blockchain/beacon/src/drand.rs

```diff
@@ -119,46 +126,24 @@ impl DrandBeacon {
         // TODO: Cache this result
         Ok(sig_match)
     }
-}
-
-#[cfg(test)]
-mod test {
-    use super::*;
 
-    async fn new_beacon() -> DrandBeacon {
-        // Current public parameters, subject to change.
-        let coeffs = [
-            hex::decode("82c279cce744450e68de98ee08f9698a01dd38f8e3be3c53f2b840fb9d09ad62a0b6b87981e179e1b14bc9a2d284c985").unwrap(),
-            hex::decode("82d51308ad346c686f81b8094551597d7b963295cbf313401a93df9baf52d5ae98a87745bee70839a4d6e65c342bd15b").unwrap(),
-            hex::decode("94eebfd53f4ba6a3b8304236400a12e73885e5a781509a5c8d41d2e8b476923d8ea6052649b3c17282f596217f96c5de").unwrap(),
-            hex::decode("8dc4231e42b4edf39e86ef1579401692480647918275da767d3e558c520d6375ad953530610fd27daf110187877a65d0").unwrap(),
-        ];
-        let dist_pub = DistPublic {
-            coefficients: coeffs,
-        };
-        DrandBeacon::new(dist_pub, 1, 25).await.unwrap()
-    }
-
-    #[async_std::test]
-    async fn construct_drand_beacon() {
-        new_beacon().await;
-    }
-
-    #[async_std::test]
-    async fn ask_and_verify_beacon_entry() {
-        let beacon = new_beacon().await;
+    async fn entry(&self, round: u64) -> Result<BeaconEntry, Box<dyn error::Error>> {
+        // TODO: Cache values into a database
+        let mut req = PublicRandRequest::new();
+        req.round = round;
+        let resp = self
+            .client
+            .public_rand(grpc::RequestOptions::new(), req)
+            .drop_metadata()
+            .await?;
 
-        let e2 = beacon.entry(2).await.unwrap();
-        let e3 = beacon.entry(3).await.unwrap();
-        assert!(beacon.verify_entry(e3, e2).unwrap());
+        Ok(BeaconEntry::new(resp.round, resp.signature))
     }
 
-    #[async_std::test]
-    async fn ask_and_verify_beacon_entry_fail() {
-        let beacon = new_beacon().await;
-
-        let e2 = beacon.entry(2).await.unwrap();
-        let e3 = beacon.entry(3).await.unwrap();
-        assert!(!beacon.verify_entry(e2, e3).unwrap());
+    fn max_beacon_round_for_epoch(&self, fil_epoch: ChainEpoch, _prev_entry: &BeaconEntry) -> u64 {
+        // TODO: sometimes the genesis time for filecoin is zero and this goes negative
+        let latest_ts = fil_epoch * self.fil_round_time + self.fil_gen_time - self.fil_round_time;
+        // TODO: self.interval has to be converted to seconds. Dont know what it is right now
+        (latest_ts - self.drand_gen_time) / self.interval
```

Comment:
I feel like this should probably be checked (a larger genesis time would just crash the node)

Reply:
Shouldn't happen in our life time. Current unix time is around: 1589569053 ~= 2^30.

Reply:
sorry, I mean drand genesis time being larger than the fil genesis time, if epoch is 0 then the line I commented on can overflow. Is there a guarantee that epoch cannot be 0? Or is drand genesis time not in unix time?

Reply:
Ah. I'm not sure actually... This is the exact math the have in Lotus though...
```
func (db *DrandBeacon) MaxBeaconRoundForEpoch(filEpoch abi.ChainEpoch, prevEntry types.BeaconEntry) uint64 {
	// TODO: sometimes the genesis time for filecoin is zero and this goes negative
	latestTs := ((uint64(filEpoch) * db.filRoundTime) + db.filGenTime) - db.filRoundTime
	dround := (latestTs - db.drandGenTime) / uint64(db.interval.Seconds())
	return dround
}

```

Reply:
Yeah, just the thing to note is go doesn't panic on overflow, we would, which is a reason against matching this (esp with their TODO comment)

Reply:
OH wow, that's wild. I never knew they dont panic on under/overflows. Let me ask Why if fil_gen_time > drand_gen_time is guaranteed. I have a feeling it is...

Reply:
Yup, confirmed!